clippy_lints/src/loops/char_indices_as_byte_indices.rs

```diff
@@ -0,0 +1,138 @@
+// Note: `String` also has methods that work with byte indices,
+// but they all take `&mut self` and aren't worth considering since the user couldn't have called
+// them while the chars iterator is live anyway.
+const BYTE_INDEX_METHODS: &[&str] = &[
+    "is_char_boundary",
+    "floor_char_boundary",
+    "ceil_char_boundary",
+    "get",
+    "index",
+    "index_mut",
+    "get_mut",
+    "get_unchecked",
+    "get_unchecked_mut",
+    "slice_unchecked",
+    "slice_mut_unchecked",
+    "split_at",
+    "split_at_mut",
+    "split_at_checked",
+    "split_at_mut_checked",
+];
+
+const CONTINUE: ControlFlow<!, ()> = ControlFlow::Continue(());
+
+pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, pat: &Pat<'_>, iterable: &Expr<'_>, body: &'tcx Expr<'tcx>) {
+    if let ExprKind::MethodCall(_, enumerate_recv, _, enumerate_span) = iterable.kind
+        && let Some(method_id) = cx.typeck_results().type_dependent_def_id(iterable.hir_id)
+        && cx.tcx.is_diagnostic_item(sym::enumerate_method, method_id)
+        && let ExprKind::MethodCall(_, chars_recv, _, chars_span) = enumerate_recv.kind
+        && let Some(method_id) = cx.typeck_results().type_dependent_def_id(enumerate_recv.hir_id)
+        && cx.tcx.is_diagnostic_item(sym::str_chars, method_id)
+    {
+        if let PatKind::Tuple([pat, _], _) = pat.kind
+            && let PatKind::Binding(_, binding_id, ..) = pat.kind
+        {
+            // Destructured iterator element `(idx, _)`, look for uses of the binding
+            for_each_expr(cx, body, |expr| {
+                if path_to_local_id(expr, binding_id) {
+                    check_index_usage(cx, expr, pat, enumerate_span, chars_span, chars_recv);
+                }
+                CONTINUE
+            });
+        } else if let PatKind::Binding(_, binding_id, ..) = pat.kind {
+            // Bound as a tuple, look for `tup.0`
+            for_each_expr(cx, body, |expr| {
+                if let ExprKind::Field(e, field) = expr.kind
+                    && path_to_local_id(e, binding_id)
+                    && field.name == sym::integer(0)
+                {
+                    check_index_usage(cx, expr, pat, enumerate_span, chars_span, chars_recv);
+                }
+                CONTINUE
+            });
+        }
+    }
+}
+
+fn check_index_usage<'tcx>(
+    cx: &LateContext<'tcx>,
+    expr: &'tcx Expr<'tcx>,
+    pat: &Pat<'_>,
+    enumerate_span: Span,
+    chars_span: Span,
+    chars_recv: &Expr<'_>,
+) {
+    let Some(parent_expr) = index_consumed_at(cx, expr) else {
+        return;
+    };
+
+    let is_string_like = |ty: Ty<'_>| ty.is_str() || is_type_lang_item(cx, ty, LangItem::String);
+    let message = match parent_expr.kind {
+        ExprKind::MethodCall(segment, recv, ..)
+            if cx.typeck_results().expr_ty_adjusted(recv).peel_refs().is_str()
+                && BYTE_INDEX_METHODS.contains(&segment.ident.name.as_str())
+                && eq_expr_value(cx, chars_recv, recv) =>
```

Comment:
I could find this check a bit suspicious in that I can't see anywhere it checks that the string hasn't been operated on or cloned (if there is then this can be ignored). In that case, the `&mut self` methods could work depending on how `eq_expr_value` works (as contrived as it is): https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1e27e7dac1f146be9a020f1c9fc822b7

Reply:
So I assume you mean this in the "false negative" sense in that your linked example could have a warning but doesn't. I suppose we could "skip" `.clone()` calls and just look at the cloned strings. Although I guess it's very unlikely for something like `string.clone().remove(idx);` in your example to exist in real code, without assigning the clone to some temporary variable first. Handling smth like
```rs
for (idx, _) in s.clone().chars().enumerate() {
  let s = s.clone();
  s.remove(idx);
}
```
could be doable I suppose, but I feel like that would complicate this lint a fair amount.


Alternatively, I do wonder if we actually need to check which string is being used at all and whether we could lint even if you use some other random/unknown string. I'm not sure if that could lead to false positives, but it would allow linting cases where you know as an invariant that two strings have the same length and then use the index into the other string or something, like
```rs
fn f(s1: &str, s2: &str) {
  assert_eq!(s1.len(), s2.len()); // invariant

  for (idx, _) in s1.chars().enumerate() {
    do_something(&s2[idx..]); // Could make sense to lint even if it isn't the same string?
  }
}
```

Reply:
> So I assume you mean this in the "false negative" sense in that your linked example could have a warning but doesn't.

I mean the opposite. iirc, when I tested it, this emitted a warning. I don't think it should. The issue I'm referencing is that if the two receivers are the same, no matter if it's just a local use or not, it'll lint. Thus you get this odd behavior.

This also means that the idea of the `String` methods being impossible to use in this way isn't necessarily correct in the comment, but it would be if this is changed